File: frontend/src/types/material.ts
/**
 * Material types for Story 13.2: Frontend My Materials Management
 *
 * TypeScript interfaces matching the backend API schemas.
 */

/**
 * Material type enumeration
 */
export type MaterialType =
  | "document"
  | "image"
  | "audio"
  | "video"
  | "url"
  | "text_note"

/**
 * Material response from API
 */
export interface Material {
  id: string
  name: string
  type: MaterialType
  file_size: number | null
  mime_type: string | null
  original_filename: string | null
  url: string | null
  text_content: string | null
  created_at: string
  updated_at: string
  download_url: string | null
}

/**
 * Storage quota information
 */
export interface StorageQuota {
  used_bytes: number
  quota_bytes: number
  used_percentage: number
  is_warning: boolean
  is_full: boolean
}

/**
 * Response when listing materials
 */
export interface MaterialListResponse {
  materials: Material[]
  total_count: number
  quota: StorageQuota
}

/**
 * Response after uploading a file or creating a material
 */
export interface UploadResponse {
  material: Material
  quota: StorageQuota
}

/**
 * Request to create a text note
 */
export interface TextNoteCreate {
  name: string
  content: string
}

/**
 * Request to update a text note
 */
export interface TextNoteUpdate {
  name?: string
  content?: string
}

/**
 * Request to create a URL link
 */
export interface UrlLinkCreate {
  name: string
  url: string
}

/**
 * Request to update material name
 */
export interface MaterialUpdate {
  name: string
}

/**
 * Upload state for tracking progress of multiple files
 */
export interface UploadingFile {
  id: string
  file: File
  progress: number
  status: "pending" | "uploading" | "complete" | "error"
  error?: string
  material?: Material
}

/**
 * Allowed file extensions by type
 */
export const ALLOWED_EXTENSIONS: Record<string, MaterialType> = {
  // Documents
  pdf: "document",
  txt: "document",
  docx: "document",
  doc: "document",
  // Images
  jpg: "image",
  jpeg: "image",
  png: "image",
  gif: "image",
  webp: "image",
  // Audio
  mp3: "audio",
  wav: "audio",
  ogg: "audio",
  m4a: "audio",
  // Video
  mp4: "video",
  webm: "video",
  mov: "video",
}

/**
 * Allowed MIME types by material type
 */
export const ALLOWED_MIME_TYPES: Record<string, MaterialType> = {
  // Documents
  "application/pdf": "document",
  "text/plain": "document",
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document":
    "document",
  "application/msword": "document",
  // Images
  "image/jpeg": "image",
  "image/png": "image",
  "image/gif": "image",
  "image/webp": "image",
  // Audio
  "audio/mpeg": "audio",
  "audio/mp3": "audio",
  "audio/wav": "audio",
  "audio/ogg": "audio",
  "audio/mp4": "audio",
  "audio/x-m4a": "audio",
  // Video
  "video/mp4": "video",
  "video/webm": "video",
  "video/quicktime": "video",
}

/**
 * Maximum file size in bytes (100MB)
 */
export const MAX_FILE_SIZE = 100 * 1024 * 1024

/**
 * Maximum text note content size in bytes (50KB)
 */
export const MAX_TEXT_NOTE_SIZE = 50 * 1024

/**
 * Display names for material types
 */
export const MATERIAL_TYPE_LABELS: Record<MaterialType, string> = {
  document: "Document",
  image: "Image",
  audio: "Audio",
  video: "Video",
  url: "URL Link",
  text_note: "Text Note",
}

/**
 * Get material type from file
 */
export function getMaterialType(file: File): MaterialType | null {
  // Try MIME type first
  if (file.type && ALLOWED_MIME_TYPES[file.type]) {
    return ALLOWED_MIME_TYPES[file.type]
  }

  // Fall back to extension
  const extension = file.name.split(".").pop()?.toLowerCase()
  if (extension && ALLOWED_EXTENSIONS[extension]) {
    return ALLOWED_EXTENSIONS[extension]
  }

  return null
}

/**
 * Validate file for upload
 */
export function validateFile(file: File): { valid: boolean; error?: string } {
  // Check file size
  if (file.size > MAX_FILE_SIZE) {
    return {
      valid: false,
      error: `File size exceeds ${MAX_FILE_SIZE / (1024 * 1024)}MB limit`,
    }
  }

  // Check file type
  const type = getMaterialType(file)
  if (!type) {
    return {
      valid: false,
      error: "File type not supported",
    }
  }

  return { valid: true }
}

File: backend/app/schemas/material.py
"""
Material API schemas for Story 13.1 - Teacher Storage Infrastructure.

Defines request/response schemas for teacher materials management.
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from app.models import MaterialType

# =============================================================================
# Request Schemas
# =============================================================================


class TextNoteCreate(BaseModel):
    """Create a text note."""

    name: str = Field(min_length=1, max_length=255)
    content: str = Field(min_length=1, max_length=51200)  # 50KB

    @field_validator("name")
    @classmethod
    def sanitize_name(cls, v: str) -> str:
        return v.strip()


class TextNoteUpdate(BaseModel):
    """Update a text note."""

    name: str | None = Field(default=None, min_length=1, max_length=255)
    content: str | None = Field(default=None, min_length=1, max_length=51200)


class UrlLinkCreate(BaseModel):
    """Create a URL link."""

    name: str = Field(min_length=1, max_length=255)
    url: str = Field(max_length=2000)

    @field_validator("url")
    @classmethod
    def validate_url(cls, v: str) -> str:
        """Basic URL validation."""
        if not v.startswith(("http://", "https://")):
            raise ValueError("URL must start with http:// or https://")
        return v


class MaterialUpdate(BaseModel):
    """Update material metadata (name only)."""

    name: str = Field(min_length=1, max_length=255)


# =============================================================================
# Response Schemas
# =============================================================================


class MaterialResponse(BaseModel):
    """Material response schema."""

    id: uuid.UUID
    name: str
    type: MaterialType
    file_size: int | None = None
    mime_type: str | None = None
    original_filename: str | None = None
    url: str | None = None  # For URL type
    text_content: str | None = None  # For text_note type
    created_at: datetime
    updated_at: datetime
    download_url: str | None = None  # Computed field for file downloads


class StorageQuotaResponse(BaseModel):
    """Storage quota information."""

    used_bytes: int
    quota_bytes: int
    used_percentage: float
    is_warning: bool  # True if > 80%
    is_full: bool  # True if >= 100%

    @property
    def available_bytes(self) -> int:
        """Calculate available bytes remaining."""
        return max(0, self.quota_bytes - self.used_bytes)


class MaterialListResponse(BaseModel):
    """List of materials with quota info."""

    materials: list[MaterialResponse]
    total_count: int
    quota: StorageQuotaResponse


class UploadResponse(BaseModel):
    """Response after file upload."""

    material: MaterialResponse
    quota: StorageQuotaResponse


class MaterialPublicResponse(BaseModel):
    """Public material info for student view (attached to assignments)."""

    id: uuid.UUID
    name: str
    type: MaterialType
    file_size: int | None = None
    mime_type: str | None = None
    url: str | None = None  # For URL type
    text_content: str | None = None  # For text_note type


class PresignedUrlResponse(BaseModel):
    """Presigned URL for direct file access."""

    url: str
    expires_in_seconds: int
    content_type: str | None = None
